### PR TITLE
apis: add APIExport/Binding/ResourceSchema types

### PR DESCRIPTION
**Pull request by sttts**
Replacing https://github.com/kcp-dev/kcp/pull/475

### CODE REVIEW

File: pkg/apis/apis/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+// fields must be set.
+type ExportReference struct {
+	// export is a reference to an exports.kcp.dev/v1alpha1 Export object uniquely
+	// identifying the APIExport to bind to.
+	//
+	// +optional
+	Export *v1alpha1.ExportBindingReference `json:"export,omitempty"`
+
+	// workspace is a reference to an APIExport in the same organization. The creator
+	// of the APIBinding needs to have access to the APIExport with the verb `bind`
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:71]: Do you prefer verb or subresource?

File: config/crds/apis.kcp.dev_apiexports.yaml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                  in the life-cycle of the APIExport. These changes have no effect
+                  on existing APIBindings, but only on newly bound ones. \n For updating
+                  existing APIBindings, use an APIDeployment keeping bound workspaces
+                  up-to-date."
+                items:
+                  type: string
+                type: array
+                x-kubernetes-list-type: set
+            type: object
+          status:
```
Comment (fabianvf) [inline config/crds/apis.kcp.dev_apiexports.yaml:55]: I could be wrong but I thought when discussing this with @ncdc there was some mention of exposing some kind of URL that points to a virtual workspace the controller can then watch in order to receive events from any workspace that imports the API

Comment (ncdc) [reply config/crds/apis.kcp.dev_apiexports.yaml:55]: We haven't gotten that far yet 😄

File: pkg/apis/apis/v1alpha1/types.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+	// in order to bind to it.
+	//
+	// +optional
+	Workspace *WorkspaceExportReference `json:"workspace,omitempty"`
+}
+
+// WorkspaceExportReference describes an API and backing implementation that are provided by an actor in the
+// same Workspace as the ServiceBinding.
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:72]:
```suggestion
// specified Workspace.
```

File: pkg/apis/apis/v1alpha1/types.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+	// +kubebuilder:validation:Required
+	// +kube:validation:MinLength=1
+	ExportName string `json:"exportName"`
+}
+
+// APIBindingPhaseType is the type of the current phase of an APIBinding.
+type APIBindingPhaseType string
+
+const (
+	APIBindingPhaseInitializing APIBindingPhaseType = "Initializing"
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:93]: This is a mismatch with the phase enum below (Binding, not Initializing). Needs to match (either way)

File: pkg/apis/apis/v1alpha1/types.go [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+	//
+	// +optional
+	Conditions conditionsv1alpha1.Conditions `json:"conditions,omitempty"`
+}
+
+// BoundAPIResource describes a bound GroupVersionResource through an APIResourceSchema of an APIExport..
+type BoundAPIResource struct {
+	// group is the group of the bound API.
+	//
+	// kubebuilder:validation:MinLength=1
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:145]: To accommodate "core v1" CRDs, we need to allow `""` for a group

File: pkg/apis/apis/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+	// +optional
+	Spec APIExportSpec `json:"spec,omitempty"`
+
+	// Status communicates the observed state.
+	//
+	// +optional
+	Status APIExportStatus `json:"status,omitempty"`
+}
+
+type APIExportSpec struct {
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:225]: Godoc (followup?)

File: pkg/apis/apis/v1alpha1/types.go [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+	//
+	// For updating existing APIBindings, use an APIDeployment keeping bound
+	// workspaces up-to-date.
+	//
+	// +optional
+	// +listType=set
+	LatestResourceSchemas []string `json:"latestResourceSchemas,omitempty"`
+}
+
+type APIExportStatus struct {
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:240]: Godoc (followup?)

File: pkg/apis/apis/v1alpha1/types.go [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec holds the desired state.
+	//
+	// +optional
+	Spec APIResourceSchemaSpec `json:"spec,omitempty"`
+}
+
+type APIResourceSchemaSpec struct {
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:283]: Godoc (followup?)

File: pkg/apis/apis/v1alpha1/types.go [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+	// group is the API group of the defined custom resource.
+	// The custom resources are served under `/apis/<group>/...`.
+	//
+	// +required
+	// +kubebuilder:validation:MinLength=1
```
Comment (ncdc) [inline pkg/apis/apis/v1alpha1/types.go:288]: To accommodate "core v1" CRDs, we need to allow `""` for a group

File: pkg/admission/apiresourceschema/admission.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	"fmt"
+	"io"
+
+	"k8s.io/apiserver/pkg/admission"
+
+	kcpadmissionhelpers "github.com/kcp-dev/kcp/pkg/admission/helpers"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+// Validate ClusterWorkspace creation and updates for valid use of spec.type, i.e. the
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/admission.go:30]: copy pasta 😄

File: pkg/admission/apiresourceschema/admission.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+type apiResourceSchemaValidation struct {
+	*admission.Handler
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&apiResourceSchemaValidation{})
+
+// Validate ensures that
+// - has a valid type
+// - has valid initializers when transitioning to initializing
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/admission.go:56]: copy pasta 😄

File: pkg/admission/apiresourceschema/admission.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+func (o *apiResourceSchemaValidation) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	if a.GetResource().GroupResource() != apisv1alpha1.Resource("apiresourceschemas") {
+		return nil
+	}
+
+	obj, err := kcpadmissionhelpers.NativeObject(a.GetObject())
+	if err != nil {
+		// nolint: nilerr
+		return nil // only work on unstructured ClusterWorkspaces
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/admission.go:65]:
```suggestion
		return nil // only work on unstructured APIResourceSchemas
```

File: pkg/admission/apiresourceschema/admission.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+	}
+	schema, ok := obj.(*apisv1alpha1.APIResourceSchema)
+	if !ok {
+		// nolint: nilerr
+		return nil // only work on unstructured ClusterWorkspaces
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/admission.go:70]:
```suggestion
		return nil // only work on unstructured APIResourceSchemas
```

File: pkg/admission/apiresourceschema/admission_test.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+			}
+			ctx := request.WithCluster(context.Background(), request.Cluster{Name: "root:org"})
+			err := o.Validate(ctx, tt.attr, nil)
+			wantErr := len(tt.expectedErrors) > 0
+			if (err != nil) != wantErr {
+				t.Fatalf("Validate() error = %v, wantErr %v", err, wantErr)
+			}
+			if err != nil {
+				t.Logf("Got admission errors: %v", err)
+				for _, expected := range tt.expectedErrors {
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/admission_test.go:147]: Would validation ever return these in a different order?

File: pkg/admission/apiresourceschema/validation.go [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+		// if the plural is not set, names validation will fail. We avoid a strange dot-dot-group message here.
+
+		expectedNamePostfix := fmt.Sprintf(".%s.%s", spec.Names.Plural, spec.Group)
+		if !strings.HasSuffix(name, expectedNamePostfix) {
+			return field.ErrorList{field.Invalid(path, name, fmt.Sprintf("must end in %s", expectedNamePostfix))}
+		}
+
+		rest := strings.TrimSuffix(name, expectedNamePostfix)
+		if !namePrefixRE.MatchString(rest) {
+			return field.ErrorList{field.Invalid(path, name, "must match patch ^[a-z]([-a-z0-9]*[a-z0-9])? in front of .<resource>.<group>")}
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/validation.go:58]: What do you mean by "must match patch"?

File: pkg/admission/apiresourceschema/validation.go [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+func ValidateAPIResourceSchemaSpec(spec *apisv1alpha1.APIResourceSchemaSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// HACK: Relax naming constraints when registering legacy schema resources through CRDs
+	// for the KCP scenario
+	if errs := utilvalidation.IsDNS1123Subdomain(spec.Group); len(errs) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), spec.Group, strings.Join(errs, ",")))
+	} else if len(strings.Split(spec.Group, ".")) < 2 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("group"), spec.Group, "should be a domain with at least one dot"))
```
Comment (ncdc) [inline pkg/admission/apiresourceschema/validation.go:73]: To accommodate "core v1" CRDs, we need to allow `""` for a group

**Review by ncdc — CHANGES_REQUESTED** [review]

File: pkg/apis/apis/v1alpha1/types.go (hunk 4 above)
Comment (sttts) [reply pkg/apis/apis/v1alpha1/types.go:93]: Binding it is

File: pkg/admission/apiresourceschema/admission_test.go (hunk 14 above)
Comment (sttts) [reply pkg/admission/apiresourceschema/admission_test.go:147]: technically not, but if it does no need to fail the test.

**Review by ncdc — APPROVED** [review]

**Pull request merged**